### PR TITLE
[Fix #4236] Add new Rails/ApplicationJob and Rails/ApplicationRecord cops

### LINKED ISSUE #4236: Suggestion: Inherit from ApplicationRecord
Hello,

I wonder if it's possible make a cop that enforces all models to inherit from ApplicationRecord instead of `ActiveRecord::Base`?

Would be a tremendous help for migrating Rails 4 -> Rails 5 applications.
And would help eliminate problems as well in code for those apps.
We ran into some weird database-fu, and had to solve it by inheriting from the correct classes for some old models.

### PR DESCRIPTION
**Pull request by tjwp**
This change implements two new cops that share a common implementation to enforce that subclasses use a specific superclass instead of the Rails base class.

### CODE REVIEW

File: lib/rubocop/cop/rails/application_job.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+      #  # bad
+      #  class Rails4Job < ActiveJob::Base
+      #    ...
+      #  end
+      class ApplicationJob < Cop
+        extend TargetRailsVersion
+
+        minimum_target_rails_version 5.0
+
+        MSG = 'Jobs should subclass ApplicationJob.'.freeze
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/application_job.rb]: Wrap the class name in backticks.

File: lib/rubocop/cop/rails/application_record.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+      #  # bad
+      #  class Rails4Model < ActiveRecord::Base
+      #    ...
+      #  end
+      class ApplicationRecord < Cop
+        extend TargetRailsVersion
+
+        minimum_target_rails_version 5.0
+
+        MSG = 'Models should subclass ApplicationRecord.'.freeze
```
Comment (bbatsov) [inline lib/rubocop/cop/rails/application_record.rb]: Same here.

**Comment by tjwp** [comment]
Backticks added

**Comment by tomhughes** [comment]
This of course completely breaks any rails app that doesn't have an ApplicationRecord class.

**Comment by mikegee** [comment]
@tomhughes this cop is only enforced for Rails 5. So the only way your app wouldn't have an ApplicationRecord class is if you deleted it from the generated code, or you upgraded to Rails 5 and didn't run [the upgrade rake task](http://guides.rubyonrails.org/upgrading_ruby_on_rails.html#the-update-task).

**Comment by tomhughes** [comment]
Yes we have upgraded over the years from 2.x and we didn't bother adding `ApplicationRecord` when updating to 5 because we didn't have any immediate need for it and it's only a bit of boilerplate.

I never run the upgrade task - to be honest I only discovered it existed recently and I can't imagine trusting an automated update on any non-trivial application. Rather what I normally do is to create a new application and use a visual diff tool to merge changes as appropriate.

**Comment by mikegee** [comment]
> Rather what I normally do is to create a new application and use a visual diff tool to merge changes as appropriate.

You might like that rake task. It lets you review all the diffs as you go.

**Pull request merged**